homeassistant/components/imap/coordinator.py

```diff
@@ -390,14 +418,25 @@ async def _async_wait_push_loop(self) -> None:
                 await asyncio.sleep(BACKOFF_TIME)
                 continue
             else:
+                self.auth_errors = 0
                 self.async_set_updated_data(number_of_messages)
+                if self.config_entry.state == ConfigEntryState.SETUP_ERROR:
+                    self.config_entry.async_set_state(
+                        self.hass,
+                        ConfigEntryState.LOADED,
+                        None,
+                    )
             try:
                 idle: asyncio.Future = await self.imap_client.idle_start()
                 await self.imap_client.wait_server_push()
                 self.imap_client.idle_done()
                 async with async_timeout.timeout(10):
                     await idle
 
+            # From python 3.11 asyncio.TimeoutError is an alias of TimeoutError
+            except asyncio.CancelledError as ex:
+                cleanup = True
```

Comment:
You could do the cleanup here but it's probably safer to do it like you have it since if cleanup raises a different exception cancelation will be effectively shielded

Another option would be a 

try: 
  Await cleanup
Finally:
  Reraise 


That would avoid the local var

Reply:
True but we don't want to raise `CancelledError` in case of an imap error or Timeout error. I could remove the finally and place it before raising and remove the local var.